packages/doenetml-worker-javascript/src/utils/externalPathResolution.ts

```diff
@@ -0,0 +1,278 @@
+import {
+    FlatPathPart,
+    FlatIndex,
+    RefResolution,
+    Resolver,
+} from "lib-doenetml-worker";
+import { PublicDoenetMLCore } from "../CoreWorker";
+
+/**
+ * Attempts to resolve the component name `name` to a componentIdx,
+ * starting the search algorithm at node `origin`.
+ *
+ * Throws an error if the name is not resolved.
+ */
+export async function resolvePathImmediatelyToNodeIdx(
+    name: string,
+    coreWorker: PublicDoenetMLCore,
+    origin = 0,
+) {
+    console.log("resolve path immediately", { name, origin });
+    // This algorithm is not a careful check of the correct form.
+    // It assumes all characters of the `name` of each path piece are word characters.
+
+    if (!coreWorker.resolvePath) {
+        return -1;
+    }
+
+    const initialMatch = name.match(/^([\w-]+)((\[\d+\])*)/);
+
+    if (!initialMatch) {
+        throw Error(`Invalid name to resolve: ${name}`);
+    }
+
+    const path: FlatPathPart[] = [
+        {
+            type: "flatPathPart",
+            name: initialMatch[1],
+            index: extractIndex(initialMatch[2]),
+        },
+    ];
+
+    let nameLeft = name.slice(initialMatch[0].length);
+
+    while (nameLeft.length > 0) {
+        const nextMatch = nameLeft.match(/^\.([\w-]+)((\[\d+\])*)/);
+        if (!nextMatch) {
+            throw Error(`Invalid name to resolve: ${name}`);
+        }
+
+        path.push({
+            type: "flatPathPart",
+            name: nextMatch[1],
+            index: extractIndex(nextMatch[2]),
+        });
+
+        nameLeft = nameLeft.slice(nextMatch[0].length);
+    }
+
+    try {
+        const resolver = coreWorker.getResolver();
+
+        const resolution = coreWorker.resolvePath(
+            resolver,
+            { path },
+            origin,
+            false,
+        );
+
+        const newResolution = await resolveAdditionalPathImmediately(
+            resolution,
+            resolver,
+            coreWorker,
+        );
+
+        if (newResolution.unresolvedPath == null) {
+            return newResolution.nodeIdx;
+        } else {
+            // If we have any unresolved path left, then the entire `path` did not match a node.
+            return -1;
```

Comment:
Sentinel values are not my favorite for return types.

Reply:
I don't have a better idea. This function is already used 40K times in tests as an index to an array!